Source/Core/Common/TraversalServer.cpp

```diff
@@ -0,0 +1,458 @@
+{
+	retry:
+	const u64 expiryTime = 30 * 1000000; // 30s
+	EvictFindResult<V> result;
+	if (map.bucket_count())
+	{
+		auto bucket = map.bucket(key);
+		auto it = map.begin(bucket);
+		for (; it != map.end(bucket); ++it)
+		{
+			if (currentTime - it->second.updateTime > expiryTime)
+			{
+				map.erase(it->first);
+				goto retry;
+			}
+			if (it->first == key)
+			{
+				if (refresh)
+					it->second.updateTime = currentTime;
+				result.found = true;
+				result.value = &it->second.value;
+				return result;
+			}
+		}
+	}
+#if DEBUG
+	printf("failed to find key '");
+	for (size_t i = 0; i < sizeof(key); i++) {
+		printf("%02x", ((u8 *) &key)[i]);
+	}
+	printf("'\n");
+#endif
+	result.found = false;
+	return result;
+}
+
+template <typename K, typename V>
+V* EvictSet(typename std::unordered_map<K, EvictEntry<V>>& map, const K& key)
+{
+	// can't use a local_iterator to emplace...
+	auto& result = map[key];
+	result.updateTime = currentTime;
+	return &result.value;
+}
+
+namespace std
+{
+	template <>
+	struct hash<TraversalHostId>
+	{
+		size_t operator()(const TraversalHostId& id) const
+		{
+			auto p = (u32*) id.data();
+			return p[0] ^ ((p[1] << 13) | (p[1] >> 19));
+		}
+	};
+}
+
+static int sock;
+static int urandomFd;
+static std::unordered_map<
+	TraversalRequestId,
+	OutgoingPacketInfo
+> outgoingPackets;
+static std::unordered_map<
+	TraversalHostId,
+	EvictEntry<TraversalInetAddress>
+> connectedClients;
+
+static TraversalInetAddress MakeInetAddress(const struct sockaddr_in6& addr)
+{
+	if (addr.sin6_family != AF_INET6)
+	{
+		fprintf(stderr, "bad sockaddr_in6\n");
+		exit(1);
+	}
+	u32* words = (u32*) addr.sin6_addr.s6_addr;
+	TraversalInetAddress result = {0};
+	if (words[0] == 0 && words[1] == 0 && words[2] == 0xffff0000)
+	{
+		result.isIPV6 = false;
+		result.address[0] = words[3];
+	}
+	else
+	{
+		result.isIPV6 = true;
+		memcpy(result.address, words, sizeof(result.address));
+	}
+	result.port = addr.sin6_port;
+	return result;
+}
+
+static struct sockaddr_in6 MakeSinAddr(const TraversalInetAddress& addr)
+{
+	struct sockaddr_in6 result;
+#ifdef SIN6_LEN
+	result.sin6_len = sizeof(result);
+#endif
+	result.sin6_family = AF_INET6;
+	result.sin6_port = addr.port;
+	result.sin6_flowinfo = 0;
+	if (addr.isIPV6)
+	{
+		memcpy(&result.sin6_addr, addr.address, 16);
+	}
+	else
+	{
+		u32* words = (u32*) result.sin6_addr.s6_addr;
+		words[0] = 0;
+		words[1] = 0;
+		words[2] = 0xffff0000;
+		words[3] = addr.address[0];
+	}
+	result.sin6_scope_id = 0;
+	return result;
+}
+
+static void GetRandomBytes(void* output, size_t size)
+{
+	static u8 bytes[8192];
+	static size_t bytesLeft = 0;
+	if (bytesLeft < size)
+	{
+		ssize_t rv = read(urandomFd, bytes, sizeof(bytes));
+		if (rv != sizeof(bytes))
+		{
+			perror("read from /dev/urandom");
+			exit(1);
+		}
+		bytesLeft = sizeof(bytes);
+	}
+	memcpy(output, bytes + (bytesLeft -= size), size);
+}
+
+static void GetRandomHostId(TraversalHostId* hostId)
```

Comment:
I wanted TraversalServer.cpp to not depend on the rest of Dolphin, since it didn't really need to, so it does not link with any other .cpp files.  Anyway, I don't think there is any reason to prefer StringFromFormat over sprintf unless the result is actually a string / variable length.